### ISSUE TITLE
bci

### ISSUE THREAD

**Issue by 9535299035** [9535299035/eeg-monitor#2]
import numpy as np
import matplotlib.pyplot as plt

from scipy.signal import butter, filtfilt, welch

from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

# ============================================================
# SETTINGS
# ============================================================

SAMPLE_RATE = 250

WINDOW_SECONDS = 2

WINDOW_SIZE = SAMPLE_RATE * WINDOW_SECONDS

# ============================================================
# FILTER
# ============================================================

def bandpass_filter(signal, low, high, fs):

    nyquist = fs / 2

    low_normalized = low / nyquist
    high_normalized = high / nyquist

    b, a = butter(
        4,
        [low_normalized, high_normalized],
        btype="band"
    )

    return filtfilt(b, a, signal)

# ============================================================
# EEG BAND POWER
# ============================================================

def band_power(signal, fs, low, high):

    frequencies, power = welch(
        signal,
        fs=fs,
        nperseg=min(len(signal), 256)
    )

    mask = (
        (frequencies >= low) &
        (frequencies <= high)
    )

    return np.trapezoid(
        power[mask],
        frequencies[mask]
    )

# ============================================================
# FEATURE EXTRACTION
# ============================================================

def extract_features(eeg_signal):

    theta = band_power(
        eeg_signal,
        SAMPLE_RATE,
        4,
        8
    )

    alpha = band_power(
        eeg_signal,
        SAMPLE_RATE,
        8,
        13
    )

    beta = band_power(
        eeg_signal,
        SAMPLE_RATE,
        13,
        30
    )

    return [
        theta,
        alpha,
        beta
    ]

# ============================================================
# BCI CLASSIFIER
# ============================================================

class BCIClassifier:

    def __init__(self):

        self.scaler = StandardScaler()

        self.model = LogisticRegression()

        self.trained = False

    def train(self, X, y):

        X = self.scaler.fit_transform(X)

        self.model.fit(
            X,
            y
        )

        self.trained = True

        print("BCI classifier trained.")

    def predict(self, features):

        if not self.trained:

            raise RuntimeError(
                "Classifier has not been trained."
            )

        features = np.asarray(
            features
        ).reshape(1, -1)

        features = self.scaler.transform(
            features
        )

        prediction = self.model.predict(
            features
        )

        probability = self.model.predict_proba(
            features
        )

        confidence = np.max(
            probability
        )

        return prediction[0], confidence

# ============================================================
# BCI COMMAND INTERPRETER
# ============================================================

def convert_to_command(
    prediction,
    confidence
):

    if confidence < 0.60:

        return "NO COMMAND"

    if prediction == 0:

        return "LEFT"

    elif prediction == 1:

        return "RIGHT"

    elif prediction == 2:

        return "SELECT"

    return "UNKNOWN"

# ============================================================
# DEMO TRAINING DATA
# ============================================================

def create_demo_dataset():

    X = []
    y = []

    for _ in range(50):

        # Simulated LEFT state
        eeg = np.random.randn(
            WINDOW_SIZE
        )

        eeg += 0.8 * np.sin(
            np.linspace(
                0,
                20 * np.pi,
                WINDOW_SIZE
            )
        )

        X.append(
            extract_features(eeg)
        )

        y.append(0)

        # Simulated RIGHT state
        eeg = np.random.randn(
            WINDOW_SIZE
        )

        eeg += 1.2 * np.sin(
            np.linspace(
                0,
                50 * np.pi,
                WINDOW_SIZE
            )
        )

        X.append(
            extract_features(eeg)
        )

        y.append(1)

        # Simulated SELECT state
        eeg = np.random.randn(
            WINDOW_SIZE
        )

        eeg += 1.5 * np.sin(
            np.linspace(
                0,
                15 * np.pi,
                WINDOW_SIZE
            )
        )

        X.append(
            extract_features(eeg)
        )

        y.append(2)

    return np.array(X), np.array(y)

# ============================================================
# LIVE BCI SIMULATION
# ============================================================

def run_demo():

    print()
    print("==============================")
    print("        EEG BCI SYSTEM")
    print("==============================")
    print()

    # Create training dataset
    X, y = create_demo_dataset()

    classifier = BCIClassifier()

    classifier.train(
        X,
        y
    )

    print()
    print("Starting BCI simulation...")
    print()

    for i in range(20):

        eeg = np.random.randn(
            WINDOW_SIZE
        )

        features = extract_features(
            eeg
        )

        prediction, confidence = classifier.predict(
            features
        )

        command = convert_to_command(
            prediction,
            confidence
        )

        print(
            f"Window {i + 1:02d} | "
            f"Command: {command:12s} | "
            f"Confidence: {confidence:.2f}"
        )

# ============================================================
# MAIN
# ============================================================

if __name__ == "__main__":

    run_demo()
LEFT
RIGHT
SELECT
NO COMMAND
          EEG HEADSET
              │
              ▼
        EEG ACQUISITION
              │
              ▼
       SIGNAL FILTERING
       ┌──────┼──────┐
       ▼      ▼      ▼
     THETA  ALPHA   BETA
       │      │      │
       └──────┼──────┘
              ▼
       FEATURE EXTRACTION
              │
              ▼
          ML MODEL
       ┌──────┼──────┐
       ▼      ▼      ▼
     LEFT   RIGHT   SELECT
              │
              ▼
        BCI APPLICATION
